src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -2851,16 +2851,19 @@ func (p *writeBatchPooledReqPool) Init() {
 
 func (p *writeBatchPooledReqPool) Get(size int) *writeBatchPooledReq {
 	cappedSize := size
-	if cappedSize > client.DefaultWriteBatchSize {
-		cappedSize = client.DefaultWriteBatchSize
+	// NB(r): Capping pooled IDs to 2x the default write batch size to account for
+	// write tagged which also has encoded tags, plus an extra one for the
+	// namespace
+	if cappedSize > (2 * client.DefaultWriteBatchSize) {
+		cappedSize = 2 * client.DefaultWriteBatchSize
 	}
 	// NB(r): Make pooled IDs plus an extra one for the namespace
 	cappedSize++
 
 	pooledReq := p.pool.Get().(*writeBatchPooledReq)
 	if cappedSize > len(pooledReq.pooledIDs) {
-		newPooledIDs := make([]writeBatchPooledReqID, 0, cappedSize)
-		newPooledIDs = append(newPooledIDs, pooledReq.pooledIDs...)
+		newPooledIDs := make([]writeBatchPooledReqID, cappedSize)
+		newPooledIDs = append(pooledReq.pooledIDs, newPooledIDs...)
```

Comment:
Looks like this would be overallocating (`pooledReq.pooledIDs + cappedSize`) instead of the intended `cappedSize`.
Also, likely 2 allocations would be happening now (unless the compiler is smart enough to elide that `newPooledIDs := make([]writeBatchPooledReqID, cappedSize)`).

What if you tried to revert to the original behaviour (allocating precisely for `cappedSize`, filling it it with a loop), assuming that [this change](https://github.com/m3db/m3/pull/3607/files#diff-8bea1e27411e7182c9db15cae2ee3cbb3d4e55d8ba57215cdf2e9231b7491abd) was the one that caused the regression?

Reply:
Hi @linasm , thank you for the valuable insight. I have updated the code, can you please review it once

Reply:
Yes, the change looks good. This pooling logic is tricky enough, I thought it might be a good idea to add some explaining comments for any future maintainers, something like this (just an optional suggestion from my side):
```go
		// 1. Allocate slice of required length in case the current length is insufficient.
		newPooledIDs := make([]writeBatchPooledReqID, cappedSize)

		// 2. Transfer existing elements into the new slice for reuse.
		copy(newPooledIDs, pooledReq.pooledIDs)

		// 3. Fill remaining part of the slice with new elements.
		for i := len(pooledReq.pooledIDs); i < len(newPooledIDs); i++ {
```

Reply:
Thank you @linasm for the suggestions, I have incorporated the comments